File: 2025-05-MAY-GFG-POTD/May 12 - Meeting Rooms III/Solution.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>

using namespace std;

class Solution {
public:
    int mostBooked(int n, vector<vector<int>>& meetings) {
        sort(meetings.begin(), meetings.end());

        // Min-heap of available rooms
        priority_queue<int, vector<int>, greater<int>> availableRooms;
        for (int i = 0; i < n; ++i) {
            availableRooms.push(i);
        }

        // Min-heap of occupied rooms (endTime, roomNumber)
        using ll = long long;
        priority_queue<pair<ll, int>, vector<pair<ll, int>>, greater<>> occupiedRooms;

        vector<int> roomUsageCount(n, 0);

        for (const auto& meeting : meetings) {
            int start = meeting[0];
            int end = meeting[1];

            // Free rooms that have completed meetings
            while (!occupiedRooms.empty() && occupiedRooms.top().first <= start) {
                availableRooms.push(occupiedRooms.top().second);
                occupiedRooms.pop();
            }

            if (!availableRooms.empty()) {
                int room = availableRooms.top(); availableRooms.pop();
                roomUsageCount[room]++;
                occupiedRooms.push({end, room});
            } else {
                auto [endTime, room] = occupiedRooms.top(); occupiedRooms.pop();
                roomUsageCount[room]++;
                ll newEnd = endTime + (end - start);
                occupiedRooms.push({newEnd, room});
            }
        }

        int maxMeetings = 0, resultRoom = 0;
        for (int i = 0; i < n; ++i) {
            if (roomUsageCount[i] > maxMeetings) {
                maxMeetings = roomUsageCount[i];
                resultRoom = i;
            }
        }

        return resultRoom;
    }
};

int main() {
    int t;
    cin >> t;
    while (t--) {
        int n, m;
        cin >> n >> m;
        vector<vector<int>> meetings(m, vector<int>(2));
        for (int i = 0; i < m; ++i) {
            cin >> meetings[i][0] >> meetings[i][1];
        }
        Solution sol;
        cout << sol.mostBooked(n, meetings) << endl;
        cout << "~" << endl;
    }
    return 0;
}

File: 2025-05-MAY-GFG-POTD/May 12 - Meeting Rooms III/Solution.java
//{ Driver Code Starts
// Initial Template for Java

import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int t = sc.nextInt(); // Number of test cases
        while (t-- > 0) {
            int n = sc.nextInt(); // Number of rooms
            int m = sc.nextInt(); // Number of meetings
            int[][] meetings = new int[m][2];
            for (int i = 0; i < m; i++) {
                meetings[i][0] = sc.nextInt(); // Start time
                meetings[i][1] = sc.nextInt(); // End time
            }
            Solution ob = new Solution();
            System.out.println(ob.mostBooked(n, meetings));
            System.out.println("~");
        }
        sc.close();
    }
}


// } Driver Code Ends

// User function Template for Java
class Solution {
    public int mostBooked(int n, int[][] meetings) {
        Arrays.sort(meetings, (a, b) -> Integer.compare(a[0], b[0]));
        
        // Min-heap of available rooms
        PriorityQueue<Integer> availableRooms = new PriorityQueue<>();
        for (int i = 0; i < n; i++) {
            availableRooms.offer(i);
        }

        // Min-heap of occupied rooms (endTime, roomNumber)
        PriorityQueue<long[]> occupiedRooms = new PriorityQueue<>((a, b) -> {
            if (a[0] == b[0]) return Long.compare(a[1], b[1]);
            return Long.compare(a[0], b[0]);
        });

        int[] roomUsageCount = new int[n];

        for (int[] meeting : meetings) {
            int start = meeting[0];
            int end = meeting[1];

            // Free rooms whose meetings have ended
            while (!occupiedRooms.isEmpty() && occupiedRooms.peek()[0] <= start) {
                availableRooms.offer((int) occupiedRooms.poll()[1]);
            }

            if (!availableRooms.isEmpty()) {
                int room = availableRooms.poll();
                roomUsageCount[room]++;
                occupiedRooms.offer(new long[]{end, room});
            } else {
                long[] earliest = occupiedRooms.poll();
                long endTime = earliest[0];
                int room = (int) earliest[1];
                roomUsageCount[room]++;
                long newEnd = endTime + (end - start);
                occupiedRooms.offer(new long[]{newEnd, room});
            }
        }

        int maxMeetings = 0;
        int resultRoom = 0;
        for (int i = 0; i < n; i++) {
            if (roomUsageCount[i] > maxMeetings) {
                maxMeetings = roomUsageCount[i];
                resultRoom = i;
            }
        }
        return resultRoom;
    }
}


//{ Driver Code Starts.
// } Driver Code Ends
